week-7 visualization, experiments, adding assumption to the camera angles, modified footprint calculation and data_model

chgenedu opened the pull request:
- added visualization and experiments
- added assumption that only one of the two camera angles may be non-zero, and that the footprint will remain rectangular
- modify the footprint related function so that the coordinates of the footprint relative to a waypoint can be used in plotting
- added function to compute waypoint speed when angle is non-zero

In src/visualization.py, added plot_photo_plan() to create plot of the flightpath and waypoint, and plot_photo_plan_with_area() to create play with flightpath with rectangles overlaying the flightpath representing areas photographed at each waypoint.

In main.ipynb, created helper functions for doing experiments, and created various experiments, for changing the following: 
- overlap
- sidelap
- height
- exposure time
- camera angle x (small and large angle)
- camera angle y (small and large angle)

In src/plan_computation.py, added waypoint_count() to count number of waypoints in each row and number of rows in a flightplan.

In src/data_model.py, in class Waypoint, add z coordinate. (Suggested in week 6 feedback.)

In the code and in the notebook, the following assumptions are added regarding camera angles:
- only one of the two camera angles (camera_angle_x_deg and camera_angle_y_deg) may be non-zero.
- The angle is small enough so that the distortion to the footprint is limited, so that the footprint will remain relatively rectangularly shaped.

About the footprint: In order to plot the photo area, I needed the coordinates of the footprint relative to each waypoint. Therefore, I reorganized the structure of compute_image_footprint_on_surface_with_angle() in src/camera_utils.py and also added compute_image_footprint_on_surface_coord().
I also added class FootprintCoord and Coordinate in src/data_model.py to store the coordinates of a footprint relative to the camera's position.

To compute waypoint speed when camera angle is non-zero, in src/plan_computation.py, I added / modified the following:
- compute_speed_during_photo_capture_with_angle()
- compute_distance_between_images_with_angle()
- closest_distance_to_footprint()

The computation of the distance to the closest point on the footprint is used to determine the fastest speed that the drone can travel in order to prevent motion blur.

chgenedu committed:
week-7 visualization, experiments, adding assumption to the camera angles, modified footprint calculation and data_model

- added visualization and experiments
- added assumption that only one of the two camera angles may be non-zero, and that the footprint will remain rectangular
- modify the footprint related function so that the coordinates of the footprint relative to a waypoint can be used in plotting
- added function to compute waypoint speed when angle is non-zero

ayushbaid commented on src/camera_utils.py:
```diff
@@ -65,6 +65,97 @@ def compute_image_footprint_on_surface(camera: Camera, distance_from_surface: fl
+
+    # Check to make sure that at most one angle is non-zero
+    # If both are non-zero, then set angle_y = 0.
+    try:
```
Curious why raise an exception and immediately catch it. If you do not want to bubble up the error, just use an if-else. Otherwise, bubble up the error to the caller.

chgenedu replied:
Modified as suggested, to simply let the error bubble up. The new version will simply raise an exception when both camera angles are non-zero, rather than setting angle_y to zero.

ayushbaid commented on src/camera_utils.py:
```diff
@@ -65,6 +65,97 @@ def compute_image_footprint_on_surface(camera: Camera, distance_from_surface: fl
+    # The coordinate on the ground where the camera aims at will be equal to
+    # math.tan(angle_x) * h
+
+    h = distance_from_surface
+    point_1 = Coordinate(math.tan(angle_x - fov_x/2)*h, math.tan(angle_y - fov_y/2)*h, -h)
+    point_2 = Coordinate(math.tan(angle_x + fov_x/2)*h, math.tan(angle_y + fov_y/2)*h, -h)
+    point_3 = Coordinate(math.tan(angle_x - fov_x/2)*h, math.tan(angle_y + fov_y/2)*h, -h)
+    point_4 = Coordinate(math.tan(angle_x + fov_x/2)*h, math.tan(angle_y - fov_y/2)*h, -h)
+    point_c = Coordinate(math.tan(angle_x)*h, math.tan(angle_y)*h, -h)
```
It would be useful if you can add Top/bottom/left/right notation to these points.

chgenedu replied:
Thanks for the suggestion. Changed as suggested.

ayushbaid commented on src/camera_utils.py:
```diff
@@ -126,3 +217,52 @@ def fov(camera: Camera) -> np.ndarray:
+
+
+# Helper function to compute the closest distance to the footprint from the camera.
+def closest_distance_to_footprint(footprint: FootprintCoord) -> float:
```
This is excellent work! You went above and beyond for ensuring that the pixel most sensitive to the blur (i.e. the closest pixel) is used here. A easy approximation would have been to just care about the center. 

This is something to highlight in the presentation as a different thing you did.

ayushbaid commented on src/plan_computation.py:
```diff
@@ -118,11 +141,17 @@ def generate_photo_plan_on_grid(camera: Camera,
+                                                                            angle_y_deg)
     # assume allowed_movment_px = 1
-    computed_speed = compute_speed_during_photo_capture(camera, dataset_spec, allowed_movement_px=1)
+    if angle_x_deg == 0 and angle_y_deg == 0:
+        computed_speed = compute_speed_during_photo_capture(camera, dataset_spec, allowed_movement_px=1)
+    else:
+        computed_speed = compute_speed_during_photo_capture_with_angle(camera,
```
I think `compute_speed_during_photo_capture_with_angle` will compute the correct speed even with both angles zero?

chgenedu replied:
Yes, but I left the original version there just in case I could not get my new version to work.

chgenedu committed:
Update main.ipynb

update to add details about experiments, and slight change of experiment code

ayushbaid reviewed (commented):
Nice work! The review of the notebook is still pending, which I will do later.

chgenedu committed:
week-7 update: added experiment, disallow two non-zero camera angles

- added experiment for varying allowed_movement_px
- modified code to enable the above experiment
- rearranged order of experiments
- modified code to disallow two non-zero camera angles in compute_image_footprint_on_surface_coord(), rather than simply forcing angle_y to be zero.

chgenedu committed:
Update main.ipynb

Added lesson learned at the end of the notebook,
and added title text for each experiment.

chgenedu committed:
Update main.ipynb

add description of some of the special functions I created

chgenedu committed:
Update main.ipynb

This version is the file with the code, but it is not executed so results are not shown.

chgenedu committed:
Create main-(code_ran_up_to_week_6).ipynb

The code in this notebook ran up to week-6 only. (This may help this file to be rendered on github.com)

chgenedu committed:
rename file

chgenedu committed:
Create main-(code_execution_for_week_7).ipynb

code executed up to week-7 experiment 4

chgenedu committed:
Update main-(code_execution_for_week_7).ipynb

Code executed up to week 7 experiment 2

chgenedu committed:
Create main-(code_executed).ipynb

main with the code executed

chgenedu committed:
code executed up to week 7 exp 2

code executed up to week 7 exp 2

chgenedu committed:
rename files

chgenedu committed:
save plotly results as html

save plotly results as html

chgenedu committed:
testing github to show images

chgenedu committed:
remove unused

chgenedu committed:
re-run main.ipynb with plotly figures saved as svg files

re-run main.ipynb with plotly figures saved as svg files

chgenedu committed:
slight cleanup

chgenedu commented:
Changed plotly figures so they are saved as static svg files and displayed in the notebook. Now the figures would show on github.

chgenedu committed:
Add a comment about interactive plot

Added a comment about how to get an interactive plot for the experiments.

ayushbaid commented on src/plan_computation.py:
```diff
@@ -99,16 +75,65 @@ def compute_speed_during_photo_capture(camera: Camera, dataset_spec: DatasetSpec
+    ### use this if both angles are zero
+    ### distance travelled during active photo capture (meters) if both angles are zero
+    # distance_without_angle = compute_ground_sampling_distance(camera, dataset_spec.height) * allowed_movement_px
+    # print(f"distance travelled during active photo capture without angle: {distance_without_angle}")
```
Nit: you can remove this now.

chgenedu replied:
Ok. Removed. Thanks.

ayushbaid commented on src/visualization.py:
```diff
@@ -2,11 +2,15 @@
 
+# import colorsys
 import plotly.graph_objects as go
+import plotly.express as px
 
-from src.data_model import Waypoint
-
+from src.data_model import Camera, DatasetSpec, Waypoint
+from src.camera_utils import compute_image_footprint_on_surface_coord
+# from src.camera_utils import compute_image_footprint_on_surface_with_angle
```
Nit: remove unused imports.

chgenedu replied:
Thanks for catching that.

ayushbaid reviewed (approved):
Excellent work!

ayushbaid commented:
@chgenedu is this ready for merge?

chgenedu committed:
Further cleanup

Remove unused code.
Re-run main.ipynb to test the programs.

chgenedu commented:
Merging.

chgenedu commented:
@ayushbaid Finished Merging. Thanks!
